subsys/net/ip/icmpv4.c

```diff
@@ -59,13 +67,130 @@ int net_icmpv4_finalize(struct net_pkt *pkt)
 	return net_pkt_set_data(pkt, &icmpv4_access);
 }
 
+static int icmpv4_parse_header_options(struct net_pkt *pkt,
+				       struct net_pkt *reply,
+				       u8_t opts_len)
+{
+	u8_t len = 0U;
+
+	/* Parse Options in EchoRequest */
+	net_pkt_cursor_init(pkt);
+
+	if (net_pkt_skip(pkt, sizeof(struct net_ipv4_hdr))) {
+		goto drop;
+	}
+
+	while (opts_len) {
+		u8_t opt_len = 0U;
+		u8_t opt_type;
+
+		/* Each option has type and length */
+		if (net_pkt_read_u8(pkt, &opt_type)) {
+			goto drop;
+		}
+
+		opts_len--;
+
+		if (!(opt_type == NET_IPV4_OPTS_EO ||
+		      opt_type == NET_IPV4_OPTS_NOP)) {
+			if (net_pkt_read_u8(pkt, &opt_len)) {
+				goto drop;
+			}
+
+			opt_len -= 2U;
+			opts_len--;
+		}
+
+		switch (opt_type) {
+		case NET_IPV4_OPTS_NOP:
+			break;
+
+		case NET_IPV4_OPTS_EO:
+			if (opts_len) {
+				/* End of options */
+				goto drop;
+			}
+
+			break;
+
+		case NET_IPV4_OPTS_RR:
+		case NET_IPV4_OPTS_TS:
+			/* TODO: Timestamp value should updated,
+			 * as per RFC 791 Internet Timestamp.
+			 * Timestamp value : 32-bit timestamp in
+			 * milliseconds since midnight UT.
+			 */
+			if (net_pkt_write_u8(reply, opt_type)) {
+				goto drop;
+			}
+
+			len++;
+
+			if (net_pkt_write_u8(reply, opt_len + 2U)) {
+				goto drop;
+			}
+
+			len++;
+
+			if (net_pkt_copy(reply, pkt, opt_len)) {
+				goto drop;
+			}
+
+			len += opt_len;
+
+			break;
+
+		default:
+			if (net_pkt_skip(pkt, opt_len)) {
+				goto drop;
+			}
+
+			break;
+		}
+
+		if (opt_len > opts_len) {
+			goto drop;
+		}
+
+		opts_len -= opt_len;
+	}
+
+	if (net_pkt_skip(pkt, NET_ICMPH_LEN)) {
+		goto drop;
+	}
+
+	/* IPv4 header should ends in 32 bit boundary */
+	if (len % 4U != 0U) {
+		u8_t i = 4U - (len % 4U);
```

Comment:
Could we replace the while loop with this?
```
net_pkt_memset(reply, NET_IPV4_OPTS_NOP, i);
len += i;
```

Reply:
Done, thanks.